File: java/calculate-sum-squares.java
class Sumthread extends Thread {
  private int start;
  private int end;
  private long result;

  public Sumthread(int pStart, int pEnd) {
    this.start = pStart;
    this.end = pEnd;
    this.result = 0;
  }

  private long sumOfSquare(int num) {
    long re = 0;
    for(int i = 1; i <= num; i++) {
      re += i*i;
    }
    return re;
  }

  public void run() {
    for(int i = this.start; i <= this.end; i++) {
      this.result += this.sumOfSquare(i);
    }
    System.out.println("Sum of squares from " + this.start + " to " + this.end + " is " + this.result + "!");
  }

  public long getResult() {
    return this.result;
  }
}

class Main {
  public static void main(String[] args) {
    int[][] ranges = {
      {100, 200},
      {201, 300},
      {301, 400},
      {401, 500},
      {501, 600},
    };
    Sumthread[] threads = new Sumthread[5];
    for(int i = 0; i < 5; i++) {
      threads[i] = new Sumthread(ranges[i][0], ranges[i][1]);
      threads[i].start();
    }
    for(int i = 0; i < 5; i++) {
      try {
        threads[i].join();
      }
      catch(InterruptedException e) {
        System.out.println(e);
      }
    }
    long total = 0;
    for(int i = 0; i < 5; i++) {
      total += threads[i].getResult();
    }
    System.out.println("Total Sum is: " + total);
  }
}

File: c/calculate-sum-squares.c
#include <stdio.h>
#include <pthread.h>

typedef struct {
  int start;
  int end;
  long result;
} ThreadValues;

long sumOfSquares(int num) {
  long re = 0;
  for(int i = 0; i <= num; i++) {
    re += i * i;
  }
  return re;
}

void* run(void* arg) {
  ThreadValues* tv = (ThreadValues*)arg;
  for(int i = tv->start; i <= tv->end; i++) {
    tv->result += sumOfSquares(i);
  }
  printf("Sum of squares from %d to %d is %ld!\n", tv->start, tv->end, tv->result);
  return NULL;
}

int main() {
  int ranges[5][2] = {
    {100, 200},
    {201, 300},
    {301, 400},
    {401, 500},
    {501, 600},
  };
  pthread_t threads[5];
  ThreadValues thread_values[5];
  for (int i = 0; i < 5; i++) {
    thread_values[i].start = ranges[i][0];
    thread_values[i].end = ranges[i][1];
    thread_values[i].result = 0;
    pthread_create(&threads[i], NULL, run, &thread_values[i]);
  }
  for (int i = 0; i < 5; i++) {
    pthread_join(threads[i], NULL);
  }
  long total = 0;
  for (int i = 0; i < 5; i++) {
    total += thread_values[i].result;
  }
  printf("Total Sum is: %ld\n", total);
  return 0;
}
